src/App/Fossa/ArchiveUploader.hs

```diff
@@ -0,0 +1,119 @@
+archiveUploadSourceUnit :: (Has Diag.Diagnostics sig m, Has (Lift IO) sig m) => Path Abs Dir -> ApiOpts -> [VendoredDependency] -> m (Maybe SourceUnit)
+archiveUploadSourceUnit baseDir apiOpts vendoredDeps = do
+  archives <- withSystemTempDir "fossa-temp" (uploadArchives apiOpts vendoredDeps baseDir)
+
+  -- archiveBuildUpload takes archives without Organization information. This orgID is appended when creating the build on the backend.
+  -- We don't care about the response here because if the build has already been queued, we get a 401 response.
+  _ <- Fossa.archiveBuildUpload apiOpts (ArchiveComponents archives)
+
+  -- The organizationID is needed to prefix each locator name. The FOSSA API automatically prefixes the locator when queuing the build
+  -- but not when reading from a source unit.
+  Fossa.Organization orgId _ <- Fossa.getOrganization apiOpts
+  let archivesWithOrganization = updateArcName (T.pack $ show orgId) <$> archives
+
+  pure $ Just $ archivesToSourceUnit archivesWithOrganization
+  where
+    updateArcName :: Text -> Archive -> Archive
+    updateArcName updateText arc = arc{archiveName = updateText <> "/" <> archiveName arc}
+
+-- archiveNoUploadSourceUnit exists for when users run `fossa analyze -o` and do not upload their source units.
+archiveNoUploadSourceUnit :: [VendoredDependency] -> Maybe SourceUnit
+archiveNoUploadSourceUnit deps = Just $ archivesToSourceUnit (unsafeVendoredToArchive <$> deps)
+
+unsafeVendoredToArchive :: VendoredDependency -> Archive
+unsafeVendoredToArchive dep = Archive (vendoredName dep) (fromMaybe "" $ vendoredVersion dep)
+
+archivesToSourceUnit :: [Archive] -> SourceUnit
+archivesToSourceUnit arcs = do
+  let build = toBuildData arcs
+      srcUnit =
+        SourceUnit
+          { sourceUnitName = "archive deps"
+          , sourceUnitManifest = "archive deps"
+          , sourceUnitType = "archive-uploaded-dependencies"
+          , sourceUnitBuild = Just build
+          , additionalData = Nothing
+          }
+  srcUnit
+
+toBuildData :: [Archive] -> SourceUnitBuild
+toBuildData deps =
+  SourceUnitBuild
+    { buildArtifact = "default"
+    , buildSucceeded = True
+    , buildImports = imports
+    , buildDependencies = map addDepList imports
+    }
+  where
+    imports = map arcToLocator deps
+
+    arcToLocator :: Archive -> Locator
+    arcToLocator arc =
+      Locator
+        { locatorFetcher = "archive"
+        , locatorProject = archiveName arc
+        , locatorRevision = Just $ archiveVersion arc
+        }
+
+    addDepList :: Locator -> SourceUnitDependency
+    addDepList loc = SourceUnitDependency loc []
+
+compressFile :: Path Abs Dir -> Path Abs Dir -> FilePath -> IO FilePath
+compressFile outputDir directory fileToTar = do
+  -- Without using `fromAbsDir` for each of these directories, the conversion
+  -- is incorrect. `show outputDir` gives an incorrect result even though it typechecks.
```

Comment:
[no action] another option here is `toFilePath :: Path a b -> FilePath` -- though using the specific `fromAbsDir` / `fromAbsFile` / etc is definitely much more descriptive

the reason `show` doesn't work here is that `show` behaves similar to the way it would for `FilePath` (or `Text` or other string types):
```hs
>>> show "foo"
\"foo\"
```

the result gets quoted